synthesizer/process/src/lib.rs

```diff
@@ -123,28 +154,103 @@ impl<N: Network> Process<N> {
         let credits_program_id = ProgramID::<N>::from_str("credits.aleo")?;
         // If the program is not 'credits.aleo', compute the program stack, and add it to the process.
         if program.id() != &credits_program_id {
-            self.add_stack(Stack::new(self, program)?);
+            self.add_stack(Arc::new(Stack::new(self, program)?))?;
         }
         Ok(())
     }
 
-    /// Adds a new stack to the process.
+    /// Adds a new stack to the LRU cache in the process.
     /// If you intend to `execute` the program, use `deploy` and `finalize_deployment` instead.
     #[inline]
-    pub fn add_stack(&mut self, stack: Stack<N>) {
-        // Add the stack to the process.
-        self.stacks.insert(*stack.program_id(), Arc::new(stack));
+    pub fn add_stack(&self, stack: Arc<Stack<N>>) -> Result<()> {
+        // Collect all direct and indirect external stacks.
+        let programs_to_add = stack.all_external_stacks();
+        // Obtain the lock on the stacks.
+        let mut stacks = self.stacks.lock();
+        // Determine which stacks still need to be added into the process.
+        let programs_to_add = programs_to_add
+            .into_iter()
+            .chain(std::iter::once((*stack.program_id(), stack))) // add the root stack.
+            .unique_by(|(id, _)|*id) // don't add duplicates.
+            .filter(|(program_id, _)| program_id != &ProgramID::<N>::from_str("credits.aleo").unwrap()) // don't add the credits.aleo stack.
+            .filter(|(program_id, _)| !stacks.contains(program_id)) // don't add stacks present in the cache.
```

Comment:
the filter should ideally be singular, or we're may end up dealing with 2 instances of the `Filter` iterator
```suggestion
            .filter(|(program_id, _)| {
                program_id != &ProgramID::<N>::from_str("credits.aleo").unwrap() // don't add the credits.aleo stack.
                && !stacks.contains(program_id)) // don't add stacks present in the cache.
            })
```